File: java/lld/booking/parking-lot/FeeStrategy.java
// Strategy: pricing is pluggable without touching the lot.
public interface FeeStrategy {
    double compute(long entryTime, long exitTime);
}

class HourlyFeeStrategy implements FeeStrategy {
    private final double ratePerHour;
    public HourlyFeeStrategy(double ratePerHour) { this.ratePerHour = ratePerHour; }

    public double compute(long entryTime, long exitTime) {
        long hours = (long) Math.ceil((exitTime - entryTime) / 3_600_000.0);
        return Math.max(1, hours) * ratePerHour;   // min 1 hour
    }
}

File: python/lld/booking/parking-lot/fee_strategy.py
import math
from abc import ABC, abstractmethod


# Strategy: pricing is pluggable without touching the lot.
class FeeStrategy(ABC):
    @abstractmethod
    def compute(self, entry_time: float, exit_time: float) -> float: ...


class HourlyFeeStrategy(FeeStrategy):
    def __init__(self, rate_per_hour: float):
        self.rate_per_hour = rate_per_hour

    def compute(self, entry_time: float, exit_time: float) -> float:
        hours = math.ceil((exit_time - entry_time) / 3_600_000)  # ms -> h
        return max(1, hours) * self.rate_per_hour                # min 1h